torchao/csrc/cpu/aten_kernels/float8_linear.cpp

```diff
@@ -0,0 +1,607 @@
+    const at::Float8_e4m3fn* B,
+    const float* scales_b,
+    int64_t M,
+    int64_t K,
+    int64_t lda,
+    int64_t ldc,
+    int64_t ldsa,
+    float* ukernel_buf,
+    at::BFloat16* dqA_buf,
+    at::BFloat16* dqB_buf) {
+  // If FP8 brgemm is not available, convert A/B to bf16 for computation
+  // Compute GEMM fp8 * fp8 -> fp32 (or bf16 * bf16 -> fp32)
+  // If per_group quant, apply scales. Otherwise, don't apply scales here
+  // Finally accumulate and store results
+#if defined(CPU_CAPABILITY_AVX512)
+  if constexpr (cpublas_can_pack) {
+#ifdef CPUBLAS_BRGEMM_F8F8F32
+    at::native::cpublas::brgemm(
+        M,
+        N,
+        K,
+        lda /*lda*/,
+        N /*ldb*/,
+        N /*ldc*/,
+        false /* add_C */,
+        A,
+        B,
+        ukernel_buf,
+        true /* is_vnni */);
+#else
+    cvt_f8e4m3_to_bf16(A, dqA_buf, M, K, lda);
+    cvt_f8e4m3_to_bf16(B, dqB_buf, K, N, N);
+    at::native::cpublas::brgemm(
+        M,
+        N,
+        K,
+        K /*lda*/,
+        N /*ldb*/,
+        N /*ldc*/,
+        false /* add_C */,
+        dqA_buf,
+        dqB_buf,
+        ukernel_buf,
+        true /* is_vnni */);
+#endif
+    _mm_prefetch(B + N * (K + 128), _MM_HINT_T0);
+    _mm_prefetch(A + K + 128, _MM_HINT_T0);
+    _accumulate_result<true, N, act_quant_mode, wei_quant_mode>(
+        C,
+        ukernel_buf,
+        scales_a,
+        scales_b,
+        M,
+        N /*ldi*/,
+        ldc,
+        ldsa);
+  } else
+#endif
+  {
+    for (int64_t i = 0; i < M; ++i) {
+      for (int64_t j = 0; j < N; ++j) {
+        float sum = 0;
+        for (int64_t k = 0; k < K; ++k) {
+          sum += ((float)A[i * lda + k] * (float)B[k * N + j]);
+        }
+        if constexpr (act_quant_mode == PER_GROUP) {
+          sum *= scales_a[i * ldsa];
+        }
+        if constexpr (wei_quant_mode == PER_GROUP) {
+          sum *= scales_b[j];
+        }
+        C[i * ldc + j] += sum;
+      }
+    }
+  }
+}
+
+template<typename out_dtype, bool cpublas_can_pack, int act_quant_mode, int wei_quant_mode>
+void _float8_linear_impl(
+    const at::Tensor& input,
+    const at::Tensor& input_scales,
+    const at::Tensor& weight,
+    const at::Tensor& weight_scales,
+    const std::optional<at::Tensor>& bias,
+    at::Tensor& output) {
+  // input shape = [..., K]
+  // input is per token quantized
+  int64_t K = input.size(-1);
+  auto input_view = input.view({-1, K});
+  int64_t M = input_view.size(0);
+
+  // weight shape = [Nc, Kc, block_k, block_n]
+  // scales shape = [Nc, G, block_n]
+  int64_t Nc = weight.size(0);
+  int64_t Kc = weight.size(1);
+  int64_t block_k = weight.size(2);
+  constexpr int64_t block_n = BLOCK_N;
+  TORCH_CHECK(weight.size(3) == block_n, "Float8 linear: unexpected weight shape");
+  int64_t N = Nc * block_n;
+  TORCH_CHECK(K == Kc * block_k, "Float8 linear: weight and input shapes mismatch");
+  auto [parallel_on_M, block_m, Mc, Mc_parallel] = get_m_blocking(M);
```

Comment:
FYI I think: 
```
  bool parallel_on_M;
  int64_t block_m, Mc, Mc_parallel;
  std::tie(parallel_on_M, block_m, Mc, Mc_parallel) = get_m_blocking(M);
```
 instead of this line fixes the error -testing internally now @Xia-Weiwen